src/privatebin.rs

```diff
@@ -165,6 +174,71 @@ impl Paste {
         let key = bs58::decode(bs58_key).into_vec()?;
         crate::crypto::decrypt_with_password(self, &key, password)
     }
+
+    /// Returns a mapping: comment.id -> decrypted_comment
+    pub fn decrypt_comments(&self, bs58_key: &str) -> PbResult<DecryptedCommentsMap> {
+        self.decrypt_comments_with_password(bs58_key, "")
+    }
+
+    /// Returns a mapping: comment.id -> decrypted_comment
+    pub fn decrypt_comments_with_password(
+        &self,
+        bs58_key: &str,
+        password: &str,
+    ) -> PbResult<DecryptedCommentsMap> {
+        let mut decrypted_comments = HashMap::new();
+        if let Some(comments) = &self.comments {
+            for comment in comments {
+                let id = comment.id.clone();
+                decrypted_comments.insert(id, comment.decrypt_with_password(bs58_key, password)?);
+            }
+        };
+        Ok(decrypted_comments)
+    }
+
+    /// Returns a mapping: comment.id -> [children comment.id]
+    pub fn comments_adjacency_map(&self) -> PbResult<CommentsAdjacencyMap> {
+        let mut comment_adjacency: CommentsAdjacencyMap = HashMap::new();
+        if let Some(comments) = &self.comments {
+            for c in comments {
+                let id = c.id.clone();
+                let parentid = c.parentid.clone();
+                comment_adjacency.entry(parentid).or_default().push(id);
+            }
+        }
+        Ok(comment_adjacency)
+    }
+
+    /// Returns formatted json trees (forest) of decrypted comments
+    pub fn comments_formatted_json_trees(
+        &self,
+        decrypted_comments: &DecryptedCommentsMap,
+        comment_adjacency: &CommentsAdjacencyMap,
+    ) -> PbResult<String> {
+        fn format_comments_below_id(
+            id: &str,
+            decrypted_comments: &DecryptedCommentsMap,
+            comment_adjacency: &CommentsAdjacencyMap,
+        ) -> serde_json::Value {
+            let formatted_children: Vec<serde_json::Value> = comment_adjacency
+                .get(id)
+                .unwrap_or(&Vec::new())
+                .iter()
+                .map(|child_id| {
+                    format_comments_below_id(child_id, decrypted_comments, comment_adjacency)
+                })
+                .collect();
+            json!({
+                id: {
```

Comment:
The current comment tree structure looks like this:

```json
 "comments": [
    {
      "e16d4261898bb4c9": {
        "comment": "Test",
        "nickname": "Mydayyy",
        "replies": [
          {
            "a9bbbe89ed779f26": {
              "comment": "Sup",
              "nickname": "Replying",
              "replies": []
            }
          }
        ]
      }
    }]
```

Is there a possibility, that an array entry has more than a single key here? If not, would this make more sense? Referring to the {id: {comment}} structure.

```json
  "comments": [
    {
      "comment": "Test",
      "id": "e16d4261898bb4c9",
      "nickname": "Mydayyy",
      "replies": [
        {
          "comment": "Sup",
          "id": "a9bbbe89ed779f26",
          "nickname": "Replying",
          "replies": []
        }
      ]
    }]
```

Which would resemble a {comment_with_id}


Here's a quick patch to make it faster to test on your side:

```
diff --git a/src/privatebin.rs b/src/privatebin.rs
index d6b4ce2..b18a0fb 100644
--- a/src/privatebin.rs
+++ b/src/privatebin.rs
@@ -229,15 +229,14 @@ impl Paste {
                 })
                 .collect();
             json!({
-                id: {
+                    "id": id,
                     "comment": decrypted_comments.get(id).unwrap_or(&DecryptedComment::default()).comment,
                     "nickname": decrypted_comments.get(id).unwrap_or(&DecryptedComment::default()).nickname,
                     "replies": formatted_children
-                }
             })
         }
         let top_level = format_comments_below_id(&self.id, decrypted_comments, comment_adjacency);
-        serde_json::to_string_pretty(&top_level[&self.id]["replies"]).map_err(|e| e.into())
+        serde_json::to_string_pretty(&top_level["replies"]).map_err(|e| e.into())
     }
 }
 
```

Reply:
I'll try this.